sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestFramework/SolutionBuilder.cs

```diff
@@ -0,0 +1,123 @@
+                new CS.CSharpParseOptions(CS.LanguageVersion.CSharp5),
+                new CS.CSharpParseOptions(CS.LanguageVersion.CSharp6),
+                new CS.CSharpParseOptions(CS.LanguageVersion.CSharp7),
+                new VB.VisualBasicParseOptions(VB.LanguageVersion.VisualBasic12),
+                new VB.VisualBasicParseOptions(VB.LanguageVersion.VisualBasic14),
+                new VB.VisualBasicParseOptions(VB.LanguageVersion.VisualBasic15)
+            );
+        }
+
+        public ProjectBuilder AddProject(string projectName, string fileExtension)
+        {
+            var language = fileExtension == GetFileExtension(AnalyzerLanguage.CSharp)
+               ? AnalyzerLanguage.CSharp
+               : AnalyzerLanguage.VisualBasic;
+
+            return AddProject(projectName, language);
+        }
+
+        public ProjectBuilder AddProject(string projectName, AnalyzerLanguage language)
+        {
+            var project = solution.AddProject(projectName, projectName, GetLanguageName(language));
+
+            return ProjectBuilder.FromProject(project)
+                .AddReference(MetadataReferenceHelper.FromFrameworkAssembly("mscorlib.dll"))
+                .AddReference(MetadataReferenceHelper.FromFrameworkAssembly("System.dll"))
+                .AddReference(MetadataReferenceHelper.FromFrameworkAssembly("System.Core.dll"))
+
+                // adding an extra file to the project
+                // this won't trigger any issues, but it keeps a reference to the original ParseOption, so
+                // if an analyzer/codefix changes the language version, Roslyn throws an ArgumentException
+                .AddSnippet(string.Empty, fileName: "ExtraEmptyFile.g." + language.GetFileExtension());
+        }
+
+        private static string GetLanguageName(AnalyzerLanguage language) =>
+            language == AnalyzerLanguage.CSharp
+                ? LanguageNames.CSharp
+                : LanguageNames.VisualBasic;
+
+        public static SolutionBuilder Create() =>
+            FromSolution(new AdhocWorkspace().CurrentSolution);
+
+        public static SolutionBuilder FromSolution(Solution solution) =>
+            new SolutionBuilder(solution);
+
+        public IReadOnlyList<Compilation> Compile(IEnumerable<ParseOptions> parseOptions = null)
+        {
+            parseOptions = parseOptions ?? DefaultParseOptions;
+
+            return solution
+                .Projects
+                .SelectMany(project => parseOptions
+                    .Where(GetParseOptionsFilter(project.Language))
+                    .Select(options => GetCompilation(project, options)))
+                .ToList()
+                .AsReadOnly();
+        }
+        private static Compilation GetCompilation(Project project, ParseOptions options) =>
+            project
+                .WithParseOptions(options)
+                .GetCompilationAsync()
+                .Result;
```

Comment:
Why are you not using `await`?

Reply:
I don't want to change all tests to asynchronous, we could do it in the future, even though it is not required.

Reply:
We cannot update the version of Roslyn we are using for unit test because of some await (and probably something else).

Reply:
Created test for that #1730